File: java/quicksort.java
import java.util.Random;

public class quicksort {
    
    public static void quicksort(int[] arr, int low, int high) {
        if (low < high) {
            int pi = partition(arr, low, high);
            quicksort(arr, low, pi - 1);
            quicksort(arr, pi + 1, high);
        }
    }
    
    public static int partition(int[] arr, int low, int high) {
        int pivot = arr[high];
        int i = low - 1;
        
        for (int j = low; j <= high - 1; j++) {
            if (arr[j] < pivot) {
                i++;
                int temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }
        int temp = arr[i + 1];
        arr[i + 1] = arr[high];
        arr[high] = temp;
        return i + 1;
    }
    
    public static int[] generateArray(int n) {
        Random random = new Random(42);
        int[] arr = new int[n];
        for (int i = 0; i < n; i++) {
            arr[i] = random.nextInt(10000);
        }
        return arr;
    }
    
    public static void main(String[] args) {
        if (args.length < 1) {
            System.out.println("Usage: quicksort <n>");
            return;
        }
        
        int n = Integer.parseInt(args[0]);
        int[] arr = generateArray(n);
        
        long start = System.currentTimeMillis();
        quicksort(arr, 0, n - 1);
        long end = System.currentTimeMillis();
        
        long timeMs = end - start;
        
        System.out.println("Java: quicksort(" + n + ") = sorted");
        System.out.println("Time: " + timeMs + "ms");
    }
}

File: zig/quicksort.zig
const std = @import("std");
const print = std.debug.print;

fn quicksort(arr: []i32, low: i32, high: i32) void {
    if (low < high) {
        const pi = partition(arr, low, high);
        quicksort(arr, low, pi - 1);
        quicksort(arr, pi + 1, high);
    }
}

fn partition(arr: []i32, low: i32, high: i32) i32 {
    const pivot = arr[@intCast(high)];
    var i = low - 1;
    
    var j = low;
    while (j <= high - 1) : (j += 1) {
        if (arr[@intCast(j)] < pivot) {
            i += 1;
            const temp = arr[@intCast(i)];
            arr[@intCast(i)] = arr[@intCast(j)];
            arr[@intCast(j)] = temp;
        }
    }
    const temp = arr[@intCast(i + 1)];
    arr[@intCast(i + 1)] = arr[@intCast(high)];
    arr[@intCast(high)] = temp;
    return i + 1;
}

fn generateArray(allocator: std.mem.Allocator, n: usize) ![]i32 {
    const arr = try allocator.alloc(i32, n);
    var seed: u32 = 42;
    
    for (arr) |*item| {
        seed = seed *% 1103515245 +% 12345;
        item.* = @intCast(seed % 10000);
    }
    return arr;
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();
    
    const args = try std.process.argsAlloc(allocator);
    defer std.process.argsFree(allocator, args);
    
    if (args.len < 2) {
        print("Usage: quicksort <n>\n", .{});
        return;
    }
    
    const n = try std.fmt.parseInt(usize, args[1], 10);
    const arr = try generateArray(allocator, n);
    defer allocator.free(arr);
    
    const start = std.time.nanoTimestamp();
    quicksort(arr, 0, @intCast(n - 1));
    const end = std.time.nanoTimestamp();
    
    const time_ms = @divFloor(end - start, 1_000_000);
    
    print("Zig: quicksort({d}) = sorted\n", .{n});
    print("Time: {d}ms\n", .{time_ms});
}